crates/matrix-sdk/src/event_cache/pagination.rs

```diff
@@ -59,17 +59,60 @@ impl RoomPagination {
     /// This automatically takes care of waiting for a pagination token from
     /// sync, if we haven't done that before.
     ///
+    /// The `until` argument is an async closure that returns a [`ControlFlow`]
+    /// to decide whether a new pagination must be run or not. It's helpful when
+    /// the server replies with e.g. a certain set of events, but we would like
+    /// more, or the event we are looking for isn't part of this set: in this
+    /// case, `until` returns [`Control::Continue`], otherwise it returns
+    /// [`ControlFlow::Break`]. `until` receives [`BackPaginationOutcome`] as
+    /// its sole argument.
+    ///
     /// # Errors
     ///
     /// It may return an error if the pagination token used during
     /// back-pagination has disappeared while we started the pagination. In
     /// that case, it's desirable to call the method again.
-    #[instrument(skip(self))]
-    pub async fn run_backwards(&self, batch_size: u16) -> Result<BackPaginationOutcome> {
+    ///
+    /// # Example
+    ///
+    /// To do a single run:
+    ///
+    /// ```rust
+    /// use std::ops::ControlFlow;
+    ///
+    /// use matrix_sdk::event_cache::{BackPaginationOutcome, RoomPagination};
+    ///
+    /// # async fn foo(room_pagination: RoomPagination) {
+    /// let result = room_pagination.run_backwards(
+    ///     42,
+    ///     |BackPaginationOutcome { events, reached_start }| async move {
+    ///         // Do something with `events` and `reached_start` maybe?
+    ///         let _ = events;
+    ///         let _ = reached_start;
+    ///
+    ///         ControlFlow::Break(())
+    ///     }
+    /// ).await;
+    /// # }
+    #[instrument(skip(self, until))]
+    pub async fn run_backwards<F, B, Fut>(&self, batch_size: u16, mut until: F) -> Result<B>
+    where
+        F: FnMut(BackPaginationOutcome) -> Fut,
+        Fut: Future<Output = ControlFlow<B, ()>>,
+    {
         loop {
-            if let Some(result) = self.run_backwards_impl(batch_size).await? {
-                return Ok(result);
+            if let Some(outcome) = self.run_backwards_impl(batch_size).await? {
+                match until(outcome).await {
```

Comment:
Unfortunately, this is not the right thing to do, because of restarts. Imagine the following situation:

- a caller would like to continue until it got 1000 events (but it uses a smaller batch size, say 100).
- the first back-pagination works and returns 100 events, the `until()` method provided will record that it's still waiting for 900 events
- a restart happens because of a timeline reset (limited flag is set)
- another back-pagination works, starting from the end of the timeline, and returns 100 events.
- the `until()` method records that we have received 100 events.

At this point, the timeline only contains 100 events, but the until() method wasn't aware of the restart so it thought that it included 200 events already.

One way to fix this would be to include a boolean parameter in the `until()` parameters' list, that indicates a timeline reset, so the caller may do appropriate action whenever that happens.

Can you include a test for this, please?

Reply:
Good eye, thanks. I've added a `TimelineHasBeenResetWhilePaginating` enum type (to avoid having a bool…), and the tests have been added/updated accordingly.